Gauge.unregister for actors

erikvanoosten opened the pull request:
Added `Gauge.unregister`.

Trait `ActorLifecycleMetricsLink` to automatically unregister gauges from restarting actors (#69).
Spec based on #70 from @scullxbones.

erikvanoosten commented:
@scullxbones, @mrmeyers99, could you please review this PR?

erikvanoosten commented:
Never mind @scullxbones , @mrmeyers99, last night I thought of a much simpler solution. I will code it tonight.

erikvanoosten commented:
I realized that keeping track of gauges could also be done in the `MetricBuilder`. After all, we have a MetricBuilder instance (embedded) per class and therefore also one per Actor. The Actor trait now simply calls the new `unregisterGauges` method on `MetricBuilder`. I had to introduce `RichAtomicReference` to make tracking the gauges thread safe.

With this approach no public API's change (in the previous `Gauge` changed) at the cost of some additional pointer house-keeping for non-actor users.

@scullxbones, @mrmeyers99, could you please review this PR?

scullxbones commented:
I like it - much, much simpler.  I guess if you wanted you could use some kind of ConcurrentMap instead of AtomicReference, but personally I think what you have is just fine.

erikvanoosten commented:
Thanks.
`ConcurrentMap` by itself doesn't work because you can not atomically empty it.

erikvanoosten commented:
@scullxbones could you please share your thoughts on the recent changes (in `master`)?
I changed the name of the trait and I added documentation. I also updated some existing documentation. Please let me know if you something that is not 100% crystal clear (or just wrong :) ).

mrmeyers99 commented:
Looks great! When will this be in a release?
